stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/channel/thread/internal/ThreadLogic.kt

```diff
@@ -16,80 +16,165 @@
 
-    /** Runs precondition for loading more messages for thread. */
-    private fun precondition(): Result<Unit> {
-        return if (mutableState.loadingOlderMessages.value) {
-            val errorMsg = "already loading messages for this thread, ignoring the load more requests."
-            logger.i { errorMsg }
-            Result(ChatError(errorMsg))
-        } else {
-            Result.success(Unit)
-        }
-    }
-
-    /** Runs side effect when a request is going to be launched. */
-    @Suppress("UnusedPrivateMember")
-    private suspend fun onRequest(messageId: String, firstId: String?, limit: Int) {
-        mutableState._loadingOlderMessages.value = true
-        // TODO implement thread replies loading from DB
-    }
-
     /** Runs side effect when a result is obtained. */
     private suspend fun onResult(result: Result<List<Message>>, limit: Int) {
         if (result.isSuccess) {
-            // Note that we don't handle offline storage for threads at the moment.
             val newMessages = result.data()
-            channelLogic.upsertMessages(newMessages)
-            mutableState._endOfOlderMessages.value = newMessages.size < limit
-            mutableState._oldestInThread.value =
-                newMessages.sortedBy { it.createdAt }.firstOrNull() ?: mutableState._oldestInThread.value
+            upsertMessages(newMessages)
+            mutableState.setEndOfOlderMessages(newMessages.size < limit)
+            mutableState.setOldestInThread(
+                newMessages.sortedBy { it.createdAt }
+                    .firstOrNull()
+                    ?: mutableState.oldestInThread.value
+            )
         }
-
-        mutableState._loadingOlderMessages.value = false
-
         result.onSuccessSuspend {
             repos.insertMessages(it)
         }
     }
 
-    override suspend fun onGetRepliesPrecondition(messageId: String, limit: Int): Result<Unit> = precondition()
+    override suspend fun onGetRepliesPrecondition(messageId: String, limit: Int): Result<Unit> {
+        return if (mutableState.loading.value) {
+            val errorMsg = "already loading messages for this thread, ignoring the load requests."
+            logger.i { errorMsg }
+            Result(ChatError(errorMsg))
+        } else {
+            val messages = repos.selectMessagesForThread(messageId, limit)
+            val parentMessage = messages.firstOrNull { it.id == messageId }
+            if (parentMessage != null) {
+                upsertMessages(messages)
+                Result.success(Unit)
+            } else {
+                val result = client.getMessage(messageId).await()
+                if (result.isSuccess) {
+                    upsertMessage(result.data())
+                    repos.insertMessage(result.data())
+                    Result.success(Unit)
+                } else {
+                    Result(result.error())
+                }
+            }
```

Comment:
I believe this codes belongs to `onGetRepliesRequest` instead of `onGetRepliesPrecondition`. Preconditions are just conditions that have to be met before a request can be made to the backend. Normally this method doesn't have any side effects and just checks for a specific state to be valid or not. 

OnGetRepliesRequest seams to be the right place to put this code because we want the database to be updated before the API request is made.